File: internal/services/strategy/engine/specmod/hash.py
"""Канонический хэш StrategySpec — совпадает с Go spechash.Hash64.

SHA-256(детерминированная сериализация proto)[:8], little-endian uint64.
Для колонки Postgres bigint приводим к signed int64.
"""

from __future__ import annotations

import hashlib

from strategy import spec_pb2


def spec_hash_u64(spec: spec_pb2.StrategySpec) -> int:
    payload = spec.SerializeToString(deterministic=True)
    digest = hashlib.sha256(payload).digest()
    return int.from_bytes(digest[:8], "little", signed=False)


def spec_hash_signed(spec: spec_pb2.StrategySpec) -> int:
    u = spec_hash_u64(spec)
    return u - 2**64 if u >= 2**63 else u

File: internal/services/strategy/manage/pkg/spechash/hash.go
// Package spechash — канонический хэш StrategySpec (дедуп прогонов и кандидатов).
package spechash

import (
	"crypto/sha256"
	"encoding/binary"
	"errors"

	strategypb "github.com/Mar1eena/trb_proto/gen/go/strategy"
	"google.golang.org/protobuf/proto"
)

// ErrSpecRequired — spec пустой.
var ErrSpecRequired = errors.New("spec обязателен")

var deterministicMarshal = proto.MarshalOptions{Deterministic: true}

// Hash64 — SHA-256(детерминированный marshal spec)[:8] little-endian как uint64.
// Совпадает по способу вычисления с indicators/manage/pkg/hash.go.
func Hash64(spec *strategypb.StrategySpec) (uint64, error) {
	if spec == nil {
		return 0, ErrSpecRequired
	}
	payload, err := deterministicMarshal.Marshal(spec)
	if err != nil {
		return 0, err
	}
	sum := sha256.Sum256(payload)
	return binary.LittleEndian.Uint64(sum[:8]), nil
}

// Signed — представление uint64-хэша для колонки Postgres bigint (без потерь).
func Signed(h uint64) int64 { return int64(h) }
